cpp/src/join/join.cu

```diff
@@ -0,0 +1,553 @@
+/**
+ * @brief  Computes the base join operation between two tables and returns the
+ * output indices of left and right table as a combined table, i.e. if full
+ * join is specified as the join type then left join is called.
+ *
+ * @throws cudf::logic_error
+ * If `left`/`right` table is empty
+ * If type mismatch between joining columns
+ *
+ * @param left  Table of left columns to join
+ * @param right Table of right  columns to join
+ * @param stream stream on which all memory allocations and copies
+ * will be performed
+ * @tparam join_kind The type of join to be performed
+ *
+ * @returns Join output indices vector pair
+ */
+/* ----------------------------------------------------------------------------*/
+template <join_kind JoinKind>
+std::pair<rmm::device_vector<size_type>,
+  rmm::device_vector<size_type>>
+get_base_join_indices(
+    table_view const& left,
+    table_view const& right,
+    cudaStream_t stream) {
+  CUDF_EXPECTS (0 != left.num_columns(), "Selected left dataset is empty");
+  CUDF_EXPECTS (0 != right.num_columns(), "Selected right dataset is empty");
+  CUDF_EXPECTS(std::equal(
+      std::cbegin(left), std::cend(left),
+      std::cbegin(right), std::cend(right),
+      [](const auto &l, const auto &r) {
+      return l.type() == r.type(); }),
+      "Mismatch in joining column data types");
+
+  constexpr join_kind BaseJoinKind = (JoinKind == join_kind::FULL_JOIN)? join_kind::LEFT_JOIN : JoinKind;
+  return get_base_hash_join_indices<BaseJoinKind>(left, right, false, stream);
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief  Combines the non common left, common left and non common right
+* columns in the correct order to form the join output table.
+*
+* @param left_noncommon_cols Columns obtained by gathering non common left
+* columns.
+* @param left_noncommon_col_indices Output locations of non common left columns
+* in the final table output
+* @param left_common_cols Columns obtained by gathering common left
+* columns.
+* @param left_common_col_indices Output locations of common left columns in the
+* final table output
+* @param right_noncommon_cols Table obtained by gathering non common right
+* columns.
+*
+* @Returns  Table containing rearranged columns.
+*/
+/* ----------------------------------------------------------------------------*/
+std::vector<std::unique_ptr<column>>
+combine_join_columns(
+    std::vector<std::unique_ptr<column>>&& left_noncommon_cols,
+    std::vector<size_type> const& left_noncommon_col_indices,
+    std::vector<std::unique_ptr<column>>&& left_common_cols,
+    std::vector<size_type> const& left_common_col_indices,
+    std::vector<std::unique_ptr<column>>&& right_noncommon_cols) {
+  std::vector<std::unique_ptr<column>> combined_cols(
+      left_noncommon_cols.size() +
+      right_noncommon_cols.size() +
+      left_common_cols.size());
+  for(size_t i = 0; i < left_noncommon_cols.size(); ++i) {
```

Comment:
Okay, how about this...

```c++
thrust::scatter(thrust::host, std::move_iterator(left_noncommon_cols.begin()), std::move_iterator(left_noncommon_cols.end()), left_noncommon_col_indices.begin(), combined_cols.begin());
```

Reply:
Sadly Thrust doesn't know what to do with r-value references since it was written pre-C++11.